pkgs/build-support/rust/lib/default.nix

```diff
@@ -5,89 +5,6 @@
 }:
 
 rec {
-  # https://doc.rust-lang.org/reference/conditional-compilation.html#target_arch
-  toTargetArch = platform:
-    /**/ if platform ? rustc.platform then platform.rustc.platform.arch
-    else if platform.isAarch32 then "arm"
-    else if platform.isMips64  then "mips64"     # never add "el" suffix
-    else if platform.isPower64 then "powerpc64"  # never add "le" suffix
-    else platform.parsed.cpu.name;
-
-  # https://doc.rust-lang.org/reference/conditional-compilation.html#target_os
-  toTargetOs = platform:
-    /**/ if platform ? rustc.platform then platform.rustc.platform.os or "none"
-    else if platform.isDarwin then "macos"
-    else platform.parsed.kernel.name;
-
-  # https://doc.rust-lang.org/reference/conditional-compilation.html#target_family
-  toTargetFamily = platform:
```

Comment:
We can't delete these without a deprecation period.

The latest release of `crate2nix` emits code that uses them.

Could you please add `lib.warn` placeholders?  We can remove them after the branch-off.

Reply:
They're not being deleted.  They're at the bottom of the file.

Reply:
```
       error: attribute 'arch' missing

       at /git/ubik/nixpkgs/pkgs/build-support/rust/lib/default.nix:70:5:

           69|   # lib.warn "`rust.${old} platform` is deprecated. Use `platform.rust.${new}` instead."
           70|     platform.rust.${new})
             |     ^
           71| {
```

Reply:
> They're at the bottom of the file.

All I see at the bottom of the file is

```nix
  toTargetArch = "arch";
```

but this doesn't work (see error above; this happens for any `Cargo.nix` produced by `crate2nix`).

It might be the case that `crate2nix` is applying `toTargetArch` to things you didn't expect it to.  It should stop doing that, but we should maintain the old behavior for one cycle with a `lib.warn` in order to let that migration happen.

Reply:
Also, I'm totally in favor of the refactoring in d51b7f8893470d998a344acfc2eefdbdf413b2cd, but it seems a little bit weird to hide it in a PR whose title is mostly about cross-compiling a C/C++ codebase (mesa).  Could you just expose whatever extra bits you need exposed for meson, and then do the cleanup in a separate PR?  I don't want waiting for `crate2nix` to switch to the new API to block `mesa` from working...

Reply:
If I revert the first and last chunks changed in `pkgs/build-support/rust/lib/default.nix` everything works again.

Could we split just those two chunks out into a separate PR in order to unblock things?  Yes I know that means there will be duplicate code.  But it will maintain the existing behavior.

Reply:
The refactoring is necessary so that we can access this data in make-derivation.nix, which is necessary to fix Mesa cross compilation.  I've been working on fixing this problem with Mesa for a year and a half.  I don't think another few days while last-minute bugs are going to make a difference.  I definitely don't want to have duplicate code, because it might start diverging, and what if something comes up and I don't have the capacity to clean it up straight away?  It's much better to get it fixed atomically.

Reply:
Fixed.  It had just got out of sync with all the different revisions this has gone through, and had to be updated to use the correct new attributes.